### ISSUE TITLE
Exception during insert using executemany() and a PL/SQL statement returns a rowcount of 0

### ISSUE THREAD

**Issue by joeltoth** [oracle/python-cx_Oracle#285]
Computronix has recently upgrade cx_Oracle to version 7.1.1 and we have noticed that the rowcount attribute is returning as 0 when a PL/SQL block with a single insert statement is passed into executemany().  

```
    dataToInsert = [
        (100, 'Jim', 'Ross', 'N', 'N'),
        (101, 'John', 'MacLean', 'N', 'N'),
        (102, 'Gary', 'Robertson', 'N', 'N'),
        (102, 'Robert', 'Pierce', 'N', 'N'),       # duplicate key
        (103, 'Bob', 'Burke', 'N', 'N'),
        (104, 'Faith', 'Hope', 'N', 'N'),
        (105, 'Rudy', 'Rudy', 'N', 'N'),
        (105, 'Charles', 'Anderson', 'N', 'N'),    # duplicate key
    ]

    #sqlStatement = "insert into ExternalContacts_t values (:1, :2, :3, :4, :5)"
    plSqlStatement = "begin insert into ExternalContacts_t values (:1, :2, :3, :4, :5); end;"

    try:
        cursor.executemany(plSqlStatement, dataToInsert)
    except cx_Oracle.DatabaseError as e:
        error, = e.args
        print("FAILED with error: %s") % error.message
        print("number of rows which succeeded (rowcount): %s") % cursor.rowcount
```

If I run the same example but instead pass a SQL statement into executemany(), the rowcount attribute returns the correct number.

This was originally fixed in cx_Oracle version 5.1.2 addessing the following item from the cx_Oracle release notes: "2. Set the row number in a cursor when executing PL/SQL blocks as requested by Robert Ritchie."

1. What is your version of Python? Is it 32-bit or 64-bit?
2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:53:40) [MSC v.1500 64 bit (AMD64)]

2. What is your cx_Oracle version?
7.1.1 & 7.1.2 tested

3. What *exact* command caused the problem (e.g. what command did you try to install with)?  Who were you logged in as?
executemany() passing in a PL/SQL block while logged in as the table owner; "sitespecific" in my case.

4. What error(s) you are seeing?
row count is being reported as 0 when i would expect it to be reporting 3 which is the number of rows that succeeded before the duplicate key exception.

5. What OS (and version) is Python executing on?
Windows 10 (1803/1809)

6. What is your version of the Oracle client (e.g. Instant Client)?  How was it installed?  Where is it installed?
Installed the Oracle client, it is not an instant client.
Versions: 11.2.0.4 , 12.1.0.2 , 12.2.0.1
Location: "C:\oracle\product\<version>\client"

7. What is your Oracle Database version?
11.2.0.4 , 12.1.0.2 , 12.2.0.1

8. Do you have a small, single Python script that immediately runs to show us the problem?
If you run the cx_Oracle Samples and change the sqlStatement to a plSqlStatement the error can be reproduced.

**Comment by joeltoth** [comment]
Please advise if this is something that is in scope to be fixed or if there is an alternative way to accomplish getting the correct rowcount when a PL/SQL statement is passed into executemany and an exception occurs for one of the rows (primary or foreign key violation, inserting a null into non null column etc).

Note: We pass PL/SQL statements to executemany() so that we can make insert calls using a package/procedure that cannot be accomplished with a single SQL statement.

**Comment by anthony-tuininga** [comment]
I took a look at it and I agree that this is something that should be fixed. Its not a common use case, but still something that works. I'll get a patch out shortly.

**Comment by anthony-tuininga** [comment]
I've just pushed a patch that corrects this issue. Let me know if that works for you. This will be included in 7.2 when it is released. Depending on what else is found I may also create a 7.1.3.

**Comment by joeltoth** [comment]
Sound good Anthony, thank you so much for your prompt reply!  

I will test the patch shortly.

**Comment by joeltoth** [comment]
This looks to have corrected the issue in 7.1.1 and 7.1.2 compiled with python 64-bit.  
I will test using 32-bit python next and let you know if I find anything.  

Thanks again for your prompt reply on this.

**Comment by joeltoth** [comment]
The compiled 32-bit .pyd is also working as expected! 

Thank you again Anthony. :)

**Comment by anthony-tuininga** [comment]
Great! Thanks for checking.

**Comment by anthony-tuininga** [comment]
FYI, just to be clear: the "row count" that is returned for PL/SQL blocks is simply equivalent to the number of blocks that have been executed. So for execute(), the value is always 1. It doesn't matter what is in the PL/SQL block either, as is demonstrated by this example (using the cx_Oracle test suite schema):

```
cursor.execute("truncate table TestTempTable")

dataToInsert = [
        (1, 'Test String 1'),
        (2, 'Test String 2'),
        (3, 'Test String 3'),
        (4, 'Test String 4')
]
cursor.executemany("""
        insert into TestTempTable (IntCol, StringCol)
        values (:1, :2)""", dataToInsert)
conn.commit()

cursor.execute("delete from TestTempTable")
print("rowcount of simple SQL statement:", cursor.rowcount)
conn.rollback()

cursor.execute("begin delete from TestTempTable; end;")
print("rowcount of PL/SQL statement:", cursor.rowcount)
conn.rollback()
```

This script shows the following result:
```
rowcount of simple SQL statement: 4
rowcount of PL/SQL statement: 1
```

For executemany(), the result is the number of iterations that have been executed successfully. As such, it isn't terribly useful, but there are occasions where it might be useful -- so I'm glad the capability was restored nonetheless!

**Comment by joeltoth** [comment]
Sounds good, thanks for the information I just have one last question regarding expected availability of the next release containing the fix.

Both patches for this bug and the @ in the password connect string have been tested and are working. While this is a working solution in development, we would like to plan for an official release of cx_Oracle for use with production systems.  You had mentioned that this fix may be rolled into 7.2.0 release or there may be a 7.1.3 release. 

Is there any timeline for when you expect the next release to come? ie next month, within 2 months, within 6 months, or is it unknown at this time?

**Comment by anthony-tuininga** [comment]
Generally speaking, the date is unknown. :-) That said, we generally try to make releases every few months at least. And since there are two issues it might be worth making cx_Oracle 7.1.3 available. I'll give it another week anyway to see if anything else crops up!

**Comment by joeltoth** [comment]
Sounds good, thank you so much for the information.  

I think we can plan for within a "few months" here but a 7.1.3 anytime sooner would be very much appreciated. :) 

Thanks again Anthony.

**Comment by joeltoth** [comment]
Good Afternoon Anthony, 

Thank you again for all of you help with resolving these issues in 7.1.2.  I have noticed that it has been roughly 3 weeks without any new issues submitted.  :) Yay! 

Is there chance that we could push for a 7.1.3 official release sometime in the near future?

**Comment by anthony-tuininga** [comment]
Yes, there is a chance. :-) I'll aim for sometime next week.

**Comment by anthony-tuininga** [comment]
I've just created a v7.1.x branch that contains what will become 7.1.3 assuming no further issues are found during release testing. It would be helpful if you could verify that it works for you, too. Thanks!

**Comment by joeltoth** [comment]
Thank you Anthony, I will try to get that tested for you over the weekend and let you know. :)

**Comment by anthony-tuininga** [comment]
The fix for this bug has been released with cx_Oracle 7.1.3.
